Allow `create` to accept a function returning a Provider

gnidan opened the pull request:
Ref: trufflesuite/truffle#348

In order to prevent all specified network providers from being instantiated / from making constant HTTP requests all the time. Instead, support wrapping provider instances in a function block to resolve instances only required for selected `--network`

gnidan committed:
Allow `create` to accept a func() -> Provider

To support `function() { return new HDWalletProvider(...) }`

gnidan commented:
Two notes:

1. This still needs tests, but truffle-provider doesn't have a test harness yet. Will add.
2. Not sure if it's worth adding underscore to the deps for this project. I can just check for functionness by hand.

gnidan committed:
Add tests for creating providers various ways

tcoulter commented on index.js:
```diff
@@ -34,7 +35,9 @@ module.exports = {
     var provider;
 
-    if (options.provider) {
+    if (options.provider && _.isFunction(options.provider)) {
```
My gut says we don't need a whole dependency for this. Can you change this line to:

```
if (options.provider && typeof options.provider == "function")) {
```

tcoulter commented on test/create.js:
```diff
@@ -0,0 +1,50 @@
+
+  beforeEach("mock HttpProvider", function() {
+    HttpProvider = Web3.providers.HttpProvider;
+    Web3.providers.HttpProvider = TestRPC.provider;
```
My gut instinct was that this monkey patch was a little gross, but then I went, "Ah, I see what you did there." I'm okay with this for now, but if you want to take the extra time and initialize an actual TestRPC server [like this](https://github.com/ethereumjs/testrpc/blob/master/test/requests.js#L930) that'd be cool too.

gnidan replied:
I'm a bit ambivalent on this because running an HTTP server is extra overhead... but it's probably actually what we want to test. Plus that means we could actually test for host/port hookup working, rather than just sending those options to the great /dev/null in the sky.

Will switch to that.

tcoulter reviewed (commented):
Things look good except for the lodash dependency, which I'd take out unless there's a more important use. Once that's fixed, the next step is to update the documentation [here](http://truffleframework.com/docs/advanced/configuration#networks). I can give you a tutorial on how its done after the merge.

gnidan committed:
Remove unnecessary lodash dependency

gnidan committed:
Switch tests to use live TestRPC server

Rather than mocking out the HttpProvider.

Also, add test for incorrect port specification.

gnidan commented:
@tcoulter took your suggestions; this is ready for your review again. Thank you!

tcoulter commented:
:+1: Just looked. Press that merge button.
